File: 1/2021.03.25/300.Longest Increasing Subsequence/python/solution.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        d = []
        for num in nums:
            if not d or num > d[-1]:
                d.append(num)
            else:
                l, r = 0, len(d) - 1
                loc = r
                while l <= r:
                    mid = l + (r - l) // 2
                    if d[mid] >= num:
                        loc = mid
                        r = mid - 1
                    else:
                        l = mid + 1
                d[loc] = num
        return len(d)

File: 1/2021.03.25/300.Longest Increasing Subsequence/java/solution.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer> res = new ArrayList<>();
        for (int num : nums) {
            if (res.isEmpty() || num > res.get(res.size() - 1)) {
                res.add(num);
            } else {
                int l = 0, r = res.size() - 1;
                int loc = r;
                while (l <= r) {
                    int mid = l + (r - l) / 2;
                    if (res.get(mid) >= num) {
                        loc = mid;
                        r = mid - 1;
                    } else {
                        l = mid + 1;
                    } 
                }
                res.set(loc, num);
            } 
        }
        return res.size();
    }
}
